hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java

```diff
@@ -136,12 +136,20 @@ public static JoinType toHibernateJoinType(int astJoinType) {
 
 		// Find tables referenced by FromReferenceNodes
 		collectReferencedTables( new ASTIterator( query ), result );
-		for (FromElement fromElement : (List<FromElement>) query.getFromClause().getFromElements()) {
+		for ( FromElement fromElement : (List<FromElement>) query.getFromClause().getFromElements() ) {
 			// For joins, we want to add the table where the association key is mapped as well as that could be a supertype that we need to join
 			String role = fromElement.getRole();
 			if ( role != null ) {
 				result.add( fromElement.getOrigin().getPropertyTableName(role.substring(role.lastIndexOf('.') + 1)) );
 			}
+			EntityPersister entityPersister = fromElement.getEntityPersister();
+			if ( entityPersister != null && entityPersister instanceof AbstractEntityPersister ) {
+				AbstractEntityPersister aep = (AbstractEntityPersister) entityPersister;
+				while ( !aep.filterFragment("", Collections.emptyMap()).isEmpty() && aep.getMappedSuperclass() != null ) {
+					Collections.addAll(result, aep.getTableNames());
+					aep = (AbstractEntityPersister) walker.getSessionFactoryHelper().findEntityPersisterByName( aep.getMappedSuperclass() );
+				}
+			}
```

Comment:
seems we can add quite some space chars to align with existing code format above